### ISSUE TITLE
bug: MV gets stuck on node failure

### ISSUE THREAD

**Issue by hzxa21** [risingwavelabs/risingwave#4553]
**Describe the bug**
I did a simple benchmark with materialized source + two-way join in a 3-node setting. After running the benchmark for ~2hrs, one compute node's heartbeat times out (althgouth the node is still alive). The streaming query is stuck starting from this point:
- CN CPU utilization drops
- Source throughput drops to zero
- I can neither query the MV nor drop the MV
![image](https://user-images.githubusercontent.com/5518566/183822285-ace7632d-ce1d-410c-801e-8809558ac788.png)

**To Reproduce**
Create a 3-node cluster and run the following query.
```
create materialized source s1 (v1 int, v2 float) with (                                                                                                                                                                                                                                                                                                             connector = 'datagen',
    fields.v1.min = '1',
    fields.v1.max  = '100000',
    fields.v2.min = '1',
    fields.v2.max = '100000',
    datagen.rows.per.second='100000',
    datagen.split.num = '12') row format json;

create materialized source s2 (v1 int, v2 float) with (                                                                                                                                                                                                                                                                                                             connector = 'datagen',
    fields.v1.min = '1',
    fields.v1.max  = '100000',
    fields.v2.min = '1',
    fields.v2.max = '100000',
    datagen.rows.per.second='100000',
    datagen.split.num = '12') row format json;

create materialized view t1 as select s1.v2 as s1v2, s2.v2 as s2v2 from s1 join s2 on s1.v1 = s2.v1;
```

**Expected behavior**
- Heartbeat should rarely times out if there is no network issue and the node is still alive.
- Failover should be able to trigger and recover when we only lose 1 out of 3 CNs in the cluster.

**Additional context**
![image](https://user-images.githubusercontent.com/5518566/183822853-377a901e-daf5-44a5-be81-a9638ae3dffc.png)

**Comment by yezizp2012** [comment]
For the recovery part, @cnissnzg has implemented migration logic in recovery when some CNs has been kicked out by meta, which means the actors in expired CNs will be migrated to new joined CNs. This requires new CNs to join. There's config option in meta named `enable_migrate` with default value `false`. I think we can remove this option and enable migration when recovery enabled. Cc @cnissnzg

**Comment by yezizp2012** [comment]
The reason why dropping mv got stuck is that meta was still under recovering, any ddl operation can't be triggered. Since scaling is still under development, we can't recovery when lose some CNs and without any new CNs joined.

**Comment by lmatz** [comment]
As scaling is almost completed, guess this issue should be already solved now?
Wonder if this case is tested in the newly added `deterministic-scale-test` or deterministic-recovery-test`.

Also weird that Github shows that this issue is mentioned in #4543, but clicking into #4543 shows no #4553

**Comment by yezizp2012** [comment]
> As scaling is almost completed, guess this issue should be already solved now? Wonder if this case is tested in the newly added `deterministic-scale-test` or deterministic-recovery-test`.
> 
> Also weird that Github shows that this issue is mentioned in #4543, but clicking into #4543 shows no #4553

In the event of loss 1 of 3 CNs in the cluster, there are two scenarios:
1. new node will join. This's already handled in recovery, the new node will take over all actors in the gone one.
2. no new node will join. Since scaling is an operation that needs to be triggered, we might need some auto scale-out strategy for this. The strategy will include a maximum wait time for nodes to join, after which auto-scale-out will be triggered Cc @BugenZhao @shanicky . Besides in a normal scale out scenario, I believe this will be well-handled on the cloud side, the operator would actively send a scale out request before going to offline the node after success.

**Comment by BugenZhao** [comment]
> new node will join. This's already handled in recovery, the new node will take over all actors in the gone one.

I believe the join of a new compute node is handled by the cloud operator. Actually, I've tested to trigger panic on the cloud and it seems to recover very well. 😄

> Heartbeat should rarely times out if there is no network issue and the node is still alive.

I believe the possibility should be significantly reduced after #4756.

**Comment by yezizp2012** [comment]
> Expected behavior

> Heartbeat should rarely times out if there is no network issue and the node is still alive.

Fixed by #4756.

> Failover should be able to trigger and recover when we only lose 1 out of 3 CNs in the cluster.

Since we've supported auto migration already, auto scale-out is quite dangerous in real production environment. for example, if the load of the left alive CNs are already relatively high, it can easily cause an avalanche. I believe we'd better handle the scenario in cloud side. I will close this issue as completed.
